File: Data Structures & Algorithms/three-integer-sum/submission-4.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    threeSum(nums) {
        nums.sort((a, b) => a - b);
        let res = new Set();

        for(let i = 0; i < nums.length; i++){
            let L = i + 1, R = nums.length - 1;

            while(L < R){
                if(R === i) R--;
                let total = nums[i] + nums[L] + nums[R];

                if(total === 0){
                    res.add(JSON.stringify([nums[i], nums[L], nums[R]].sort((a, b) => a - b)));
                }
                
                if(total > 0) R--;

                else{
                    L++;
                }
            }
        }
        return [...res].map((x) => JSON.parse(x));
    }
}

File: Data Structures & Algorithms/subarray-sum-equals-k/submission-4.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        res = 0

        prefixSums = {0 : 1}
        #prefix : times it occurs

        total = 0

        i = 0

        while i < len(nums):
            total += nums[i]

            if total - k in prefixSums:
                res += prefixSums[total - k]

            if total in prefixSums:
                prefixSums[total] += 1
            else:
                prefixSums[total] = 1
            
            i += 1
        return res

        
        
        